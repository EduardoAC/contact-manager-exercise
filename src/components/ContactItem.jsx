import React from 'react';
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom';

const ContactItem = (props) => {
  const {
    id,
    firstName,
    lastName,
    phoneNumber,
    address,
    email,
  } = props;
  return (
    <tr className="contact__item">
      <td>{id}</td>
      <td>{firstName}</td>
      <td>{lastName}</td>
      <td>{address}</td>
      <td>{phoneNumber}</td>
      <td>{email}</td>
      <td>
        <Link to={`/${id}/edit`} style={{ marginRight: '5px' }}>Edit</Link>
        <Link to={`/${id}/delete`}>Delete</Link>
      </td>
    </tr>
  )
};

ContactItem.propTypes = {
  id: PropTypes.string.isRequired,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  phoneNumber: PropTypes.string.isRequired,
  address: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
};

export default ContactItem;
