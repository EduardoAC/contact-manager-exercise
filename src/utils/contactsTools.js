export function getContactId(props) {
  return props.match && props.match.params && props.match.params.id
};

export function getContactFromState(state, props) {
  const id = getContactId(props);
  let contact = undefined;
  let index = -1;
  if (id) {
    index = state.contacts.index[id];
  }
  if (index >= 0) {
    contact = state.contacts.values[index];
  }
  return contact;
}

export function removeContact(array, index) {
  let newContacts = array.slice();
  newContacts.splice(index, 1);
  const newContactIndex = newContacts.reduce((a, c, i) => {
    a[c.id] = i;
    return a;
  }, {});
  return { index: newContactIndex, values: newContacts};
}
