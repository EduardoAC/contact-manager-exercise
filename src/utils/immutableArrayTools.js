export function insertItem(array, item) {
    return [...array, item];
}

export function updateObjectInArray(array, item, index) {
    return array.map( (el, key) => {
        if(key !== index) {
            // This isn't the item we care about - keep it as-is
            return el;
        }
        // Otherwise, this is the one we want - return an updated value
        return {
            ...el,
            ...item
        };
    });
}
