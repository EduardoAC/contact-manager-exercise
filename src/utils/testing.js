import faker from 'faker';
import guid from './guid';

const getFakeContact = () => ({
  id: guid(),
  firstName: faker.name.firstName(),
  lastName: faker.name.lastName(),
  phoneNumber: faker.phone.phoneNumber(),
  address: faker.address.streetAddress(true),
  email: faker.internet.email(),
});
const generateMockContacts = (size) => (
  Array.from(new Array(size), getFakeContact)
);

export { getFakeContact, generateMockContacts };
