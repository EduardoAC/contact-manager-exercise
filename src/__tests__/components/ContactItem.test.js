import React from 'react';
import { shallow, mount } from 'enzyme';
import { Link } from 'react-router-dom';

import ContactItem from '../../components/ContactItem';
import { getFakeContact } from '../../utils/testing';

describe('<ContactItem />', () => {
  let wrapper
  it('Component renders correctly', () => {
    const wrapper = shallow(<ContactItem {...getFakeContact()} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('contains an edit link to modify contact', () => {
    const contact = getFakeContact();
    const wrapper = shallow(<ContactItem {...contact} />);
    expect(wrapper.find(Link).first().props().to).toBe(`/${contact.id}/edit`);
  });

  it('contains an delete link to modify contact', () => {
    const contact = getFakeContact();
    const wrapper = shallow(<ContactItem {...contact} />);
    expect(wrapper.find(Link).at(1).props().to).toBe(`/${contact.id}/delete`);
  });
});
