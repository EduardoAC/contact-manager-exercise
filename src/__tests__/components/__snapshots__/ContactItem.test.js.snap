// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactItem /> Component redenrs correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <tr
    className="contact__item"
>
    <td>
        98a59da4-9781-a957-b446-f27cb63e3b52
    </td>
    <td>
        Denis
    </td>
    <td>
        Becker
    </td>
    <td>
        0313 Aurelio Creek Apt. 778
    </td>
    <td>
        (642) 649-8274 x570
    </td>
    <td>
        Thad_Kerluke@yahoo.com
    </td>
    <td>
        <Link
            replace={false}
            style={
                Object {
                    "marginRight": "5px",
                  }
            }
            to="/98a59da4-9781-a957-b446-f27cb63e3b52/edit"
        >
            Edit
        </Link>
        <Link
            replace={false}
            to="/98a59da4-9781-a957-b446-f27cb63e3b52/delete"
        >
            Delete
        </Link>
    </td>
</tr>,
  "nodes": Array [
    <tr
      className="contact__item"
>
      <td>
            98a59da4-9781-a957-b446-f27cb63e3b52
      </td>
      <td>
            Denis
      </td>
      <td>
            Becker
      </td>
      <td>
            0313 Aurelio Creek Apt. 778
      </td>
      <td>
            (642) 649-8274 x570
      </td>
      <td>
            Thad_Kerluke@yahoo.com
      </td>
      <td>
            <Link
                  replace={false}
                  style={
                        Object {
                              "marginRight": "5px",
                            }
                  }
                  to="/98a59da4-9781-a957-b446-f27cb63e3b52/edit"
            >
                  Edit
            </Link>
            <Link
                  replace={false}
                  to="/98a59da4-9781-a957-b446-f27cb63e3b52/delete"
            >
                  Delete
            </Link>
      </td>
</tr>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ContactItem
        address="0313 Aurelio Creek Apt. 778"
        email="Thad_Kerluke@yahoo.com"
        firstName="Denis"
        id="98a59da4-9781-a957-b446-f27cb63e3b52"
        lastName="Becker"
        phoneNumber="(642) 649-8274 x570"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "address": "0313 Aurelio Creek Apt. 778",
          "email": "Thad_Kerluke@yahoo.com",
          "firstName": "Denis",
          "id": "98a59da4-9781-a957-b446-f27cb63e3b52",
          "lastName": "Becker",
          "phoneNumber": "(642) 649-8274 x570",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <tr
          className="contact__item"
>
          <td>
                    98a59da4-9781-a957-b446-f27cb63e3b52
          </td>
          <td>
                    Denis
          </td>
          <td>
                    Becker
          </td>
          <td>
                    0313 Aurelio Creek Apt. 778
          </td>
          <td>
                    (642) 649-8274 x570
          </td>
          <td>
                    Thad_Kerluke@yahoo.com
          </td>
          <td>
                    <Link
                              replace={false}
                              style={
                                        Object {
                                                  "marginRight": "5px",
                                                }
                              }
                              to="/98a59da4-9781-a957-b446-f27cb63e3b52/edit"
                    >
                              Edit
                    </Link>
                    <Link
                              replace={false}
                              to="/98a59da4-9781-a957-b446-f27cb63e3b52/delete"
                    >
                              Delete
                    </Link>
          </td>
</tr>,
        "_debugID": 2,
        "_renderedOutput": <tr
          className="contact__item"
>
          <td>
                    98a59da4-9781-a957-b446-f27cb63e3b52
          </td>
          <td>
                    Denis
          </td>
          <td>
                    Becker
          </td>
          <td>
                    0313 Aurelio Creek Apt. 778
          </td>
          <td>
                    (642) 649-8274 x570
          </td>
          <td>
                    Thad_Kerluke@yahoo.com
          </td>
          <td>
                    <Link
                              replace={false}
                              style={
                                        Object {
                                                  "marginRight": "5px",
                                                }
                              }
                              to="/98a59da4-9781-a957-b446-f27cb63e3b52/edit"
                    >
                              Edit
                    </Link>
                    <Link
                              replace={false}
                              to="/98a59da4-9781-a957-b446-f27cb63e3b52/delete"
                    >
                              Delete
                    </Link>
          </td>
</tr>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ContactItem
    address="0313 Aurelio Creek Apt. 778"
    email="Thad_Kerluke@yahoo.com"
    firstName="Denis"
    id="98a59da4-9781-a957-b446-f27cb63e3b52"
    lastName="Becker"
    phoneNumber="(642) 649-8274 x570"
/>,
}
`;

exports[`<ContactItem /> Component renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <tr
    className="contact__item"
>
    <td>
        def433df-9c9a-eaff-cc91-66ff0dd4505b
    </td>
    <td>
        Vivien
    </td>
    <td>
        Romaguera
    </td>
    <td>
        7699 Abernathy Burg Apt. 462
    </td>
    <td>
        081-301-5588 x534
    </td>
    <td>
        Serenity22@hotmail.com
    </td>
    <td>
        <Link
            replace={false}
            style={
                Object {
                    "marginRight": "5px",
                  }
            }
            to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/edit"
        >
            Edit
        </Link>
        <Link
            replace={false}
            to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/delete"
        >
            Delete
        </Link>
    </td>
</tr>,
  "nodes": Array [
    <tr
      className="contact__item"
>
      <td>
            def433df-9c9a-eaff-cc91-66ff0dd4505b
      </td>
      <td>
            Vivien
      </td>
      <td>
            Romaguera
      </td>
      <td>
            7699 Abernathy Burg Apt. 462
      </td>
      <td>
            081-301-5588 x534
      </td>
      <td>
            Serenity22@hotmail.com
      </td>
      <td>
            <Link
                  replace={false}
                  style={
                        Object {
                              "marginRight": "5px",
                            }
                  }
                  to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/edit"
            >
                  Edit
            </Link>
            <Link
                  replace={false}
                  to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/delete"
            >
                  Delete
            </Link>
      </td>
</tr>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ContactItem
        address="7699 Abernathy Burg Apt. 462"
        email="Serenity22@hotmail.com"
        firstName="Vivien"
        id="def433df-9c9a-eaff-cc91-66ff0dd4505b"
        lastName="Romaguera"
        phoneNumber="081-301-5588 x534"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "address": "7699 Abernathy Burg Apt. 462",
          "email": "Serenity22@hotmail.com",
          "firstName": "Vivien",
          "id": "def433df-9c9a-eaff-cc91-66ff0dd4505b",
          "lastName": "Romaguera",
          "phoneNumber": "081-301-5588 x534",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <tr
          className="contact__item"
>
          <td>
                    def433df-9c9a-eaff-cc91-66ff0dd4505b
          </td>
          <td>
                    Vivien
          </td>
          <td>
                    Romaguera
          </td>
          <td>
                    7699 Abernathy Burg Apt. 462
          </td>
          <td>
                    081-301-5588 x534
          </td>
          <td>
                    Serenity22@hotmail.com
          </td>
          <td>
                    <Link
                              replace={false}
                              style={
                                        Object {
                                                  "marginRight": "5px",
                                                }
                              }
                              to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/edit"
                    >
                              Edit
                    </Link>
                    <Link
                              replace={false}
                              to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/delete"
                    >
                              Delete
                    </Link>
          </td>
</tr>,
        "_debugID": 2,
        "_renderedOutput": <tr
          className="contact__item"
>
          <td>
                    def433df-9c9a-eaff-cc91-66ff0dd4505b
          </td>
          <td>
                    Vivien
          </td>
          <td>
                    Romaguera
          </td>
          <td>
                    7699 Abernathy Burg Apt. 462
          </td>
          <td>
                    081-301-5588 x534
          </td>
          <td>
                    Serenity22@hotmail.com
          </td>
          <td>
                    <Link
                              replace={false}
                              style={
                                        Object {
                                                  "marginRight": "5px",
                                                }
                              }
                              to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/edit"
                    >
                              Edit
                    </Link>
                    <Link
                              replace={false}
                              to="/def433df-9c9a-eaff-cc91-66ff0dd4505b/delete"
                    >
                              Delete
                    </Link>
          </td>
</tr>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ContactItem
    address="7699 Abernathy Burg Apt. 462"
    email="Serenity22@hotmail.com"
    firstName="Vivien"
    id="def433df-9c9a-eaff-cc91-66ff0dd4505b"
    lastName="Romaguera"
    phoneNumber="081-301-5588 x534"
/>,
}
`;
